### PR TITLE
avoid resizing if capacity is enough when (de)compressing

### LINKED ISSUE #6276: A better way to resize the buffer for the snappy encode/decode
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

In my use case where reading data from a big parquet file on disk, the snappy decoding costs too much cpu resources in my flamegraph. And digging into the codebase, I find that `resize` is used to initialize the buffer for the decoding destination buffer, but `resize` shows a bad performance when the expected size is large enough.

Here is the code location about the `resize`:

https://github.com/apache/arrow-rs/blob/25d39c13fc8937e86373821093d063202484ac00/parquet/src/compression.rs#L211-L226

And here is an example showing the bad performance of `resize` a vector:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5321aaeda859ea9f664ae7952ade2fb6

And the example's output says:
```
init with macro, num_elems:10000, cost:7.74µs
init with resize, num_elems:20000, cost:196.694µs
init with set_len, num_elems:30000, cost:5.171µs
```

It proves `resize` a vector to 10000 costs too much more time than `vec!` macro or `set_len`.

**Describe the solution you'd like**

In the example above, `set_len` shows a good performance, so I guess `resize` can be replaced with `set_len` if the capacity is enough:
```rust
let new_len = offset + len;
if output_buf.capacity() >= new_len {
    unsafe {
        output_buf.set_len(new_len);
    }
} else {
    output_buf.resize(new_len, 0);
}
```

Although `unsafe` block is introduced here, it is actually safe considering:
1. the element of the vector is byte (no worry about the destruction);
2. the uninitialized content will be overwritten immediately.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by ShiKaiWi**
# Which issue does this PR close?

Closes #6276

# Rationale for this change
Currently, the buffer used for the (de)compression is resized to the expected size with `resizing`. However, the `resize` will do initialization work which is unnecessary for the (de)compression because the contents will be overwritten immediately.

# What changes are included in this PR?
Introduces a new safe way to do resize without initialization on the `Vec<u8>`, and use it to resize the output buffer's size before (de)compression.

# Are there any user-facing changes?
None.

### CODE REVIEW

File: parquet/src/util/vec_util.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+/// Resize the `buf` to a new len `n` without initialization.
+///
+/// Replacing `resize` on the `buf` with `reserve` and `set_len` can skip the initialization
+/// cost for a good performance.
+/// And the `set_len` here is safe because the element of the vector is byte whose destruction
+/// does nothing.
+// TODO: remove this clippy allow lint if it is used by a module without feature gate.
```
Comment (Xuanwo) [inline parquet/src/util/vec_util.rs:24]: How about hiding it within the same feature gate?

File: parquet/src/util/vec_util.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+#[allow(dead_code)]
+pub fn resize_buffer_without_init(buf: &mut Vec<u8>, n: usize) {
+    if n > buf.capacity() {
+        buf.reserve(n - buf.len());
+    }
+    unsafe { buf.set_len(n) };
```
Comment (Xuanwo) [inline parquet/src/util/vec_util.rs:30]: How about adding a safety comment for `set_len`

File: parquet/src/compression.rs [hunk 3]
```diff
@@ -150,35 +150,47 @@ pub fn create_codec(codec: CodecType, _options: &CodecOptions) -> Result<Option<
...
+                "Disabled feature at compile time: zstd".into(),
+            ))
+        }
         CodecType::LZ4_RAW => {
             #[cfg(any(feature = "lz4", test))]
             return Ok(Some(Box::new(LZ4RawCodec::new())));
-            Err(ParquetError::General("Disabled feature at compile time: lz4".into()))
-        },
+            Err(ParquetError::General(
+                "Disabled feature at compile time: lz4".into(),
```
Comment (Xuanwo) [inline parquet/src/compression.rs:191]: Looks like unexpected format changes.

**Review by Xuanwo — COMMENTED** [review]
Thanks a lot for this change. Would you like to provide a bench of this change?

**Comment by ShiKaiWi** [comment]
> Thanks a lot for this change. Would you like to provide a bench of this change?

Sure.

**Review by tustvold — CHANGES_REQUESTED** [review]
This is UB as per the safety contract on Vec. I suggest you take a look at MaybeUninit for how to do this sort of thing safely

**Comment by alamb** [comment]
I also think if we want to add an optimization such as this, it needs to include some benchmark results showing the improved performance is worth the added code.

**Comment by ShiKaiWi** [comment]
> This is UB as per the safety contract on Vec::set_len. I suggest you take a look at MaybeUninit for how to do this sort of thing safely, and also why what you're doing here is UB

@tustvold Thanks a lot for your suggestions.

>  I suggest you take a look at MaybeUninit for how to do this sort of thing safely

The `Vec<u8>` will be used by the snnapy (de)compression, so I guess introducing `MaybeUninit` is not allowed if we try to change the signature of the buffer.

> why what you're doing here is UB

Actually, this is UB here for the memory between `buf[old_len..new_len]` is uninitialized, but the tests in the CI works fine. And I guess the reason for this is that the uninitialized content won't be read before writes (or initialization) in the (de)compression process.

And I find something interesting in the [example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5321aaeda859ea9f664ae7952ade2fb6) mentioned at #6276, `vec!` macro also performs much better than `resize` (20x faster), so I guess there are some magic in the `vec!` macro.

After digging into the std source code about `vec!` macro, I find a fast path for the initialization of `Vec<u8>` and `Vec<i8>`. And the two code snippets show the fast path:

https://github.com/rust-lang/rust/blob/982c6f8721416431ec62bb0b9105c0578a9fc603/library/alloc/src/vec/mod.rs#L2752-L2754
```rust
pub fn from_elem<T: Clone>(elem: T, n: usize) -> Vec<T> {
    <T as SpecFromElem>::from_elem(elem, n, Global)
}
```

https://github.com/rust-lang/rust/blob/982c6f8721416431ec62bb0b9105c0578a9fc603/library/alloc/src/vec/spec_from_elem.rs#L47-L60
```rust
impl SpecFromElem for u8 {
    #[inline]
    fn from_elem<A: Allocator>(elem: u8, n: usize, alloc: A) -> Vec<u8, A> {
        if elem == 0 {
            return Vec { buf: RawVec::with_capacity_zeroed_in(n, alloc), len: n };
        }
        let mut v = Vec::with_capacity_in(n, alloc);
        unsafe {
            ptr::write_bytes(v.as_mut_ptr(), elem, n);
            v.set_len(n);
        }
        v
    }
}
```

And a new [example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a407cda52e30f36bf730be2709c7bf5e) shows using the `ptr::write_bytes` + `Vec::set_len` performs much better than `resize` too.

So how about using `ptr::write_bytes` to do initialization work before calling `Vec::set_len`? @tustvold Looking forward to your further suggestions!

**Comment by ShiKaiWi** [comment]
> I also think if we want to add an optimization such as this, it needs to include some benchmark results showing the improved performance is worth the added code.

Actually, I've provided a simple [example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5321aaeda859ea9f664ae7952ade2fb6) in #6276 to show the rationale for this changeset. Anyway, I'll add a benchmark in the codebase if the new solution is agreed. 😊

**Comment by alamb** [comment]
> Anyway, I'll add a benchmark in the codebase if the new solution is agreed. 😊

Sounds good!

**Comment by etseidl** [comment]
I agree that results from the compression bench are needed. Running your test code in release mode speeds things up dramatically.
```
Debug
init with macro, num_elems:10000, cost:7.771µs
init with resize, num_elems:20000, cost:130.553µs
init with set_len, num_elems:30000, cost:7.661µs

Release
init with macro, num_elems:10000, cost:100ns
init with resize, num_elems:20000, cost:7.49µs
init with set_len, num_elems:30000, cost:80ns
```
`resize` is still much slower, but is 20X faster with optimization enabled. I'd be curious to see how overall throughput through the decompressor is affected by this change.

FWIW I tried replacing another [use](https://github.com/apache/arrow-rs/blob/30db5dce9ca0996457063f1b5308076a6372c438/parquet/src/arrow/array_reader/fixed_len_byte_array.rs#L467) of `resize` with the `reserve/write_bytes/set_len` proposed here and saw a modest (2-4%) speedup in decoding times.
```
group                                                                                                   new_resize                             to_prim
-----                                                                                                   ----------                             -------
arrow_array_reader/BYTE_STREAM_SPLIT/Decimal128Array/byte_stream_split encoded, mandatory, no NULLs     1.00    411.7±2.05µs        ? ?/sec    1.04    429.1±7.35µs        ? ?/sec
arrow_array_reader/BYTE_STREAM_SPLIT/Decimal128Array/byte_stream_split encoded, optional, half NULLs    1.00    523.3±2.65µs        ? ?/sec    1.02    532.7±4.72µs        ? ?/sec
arrow_array_reader/BYTE_STREAM_SPLIT/Decimal128Array/byte_stream_split encoded, optional, no NULLs      1.00    414.0±3.41µs        ? ?/sec    1.04    428.6±6.21µs        ? ?/sec
arrow_array_reader/BYTE_STREAM_SPLIT/Float16Array/byte_stream_split encoded, mandatory, no NULLs        1.00     52.1±0.26µs        ? ?/sec    1.04     54.1±0.78µs        ? ?/sec
arrow_array_reader/BYTE_STREAM_SPLIT/Float16Array/byte_stream_split encoded, optional, half NULLs       1.00    109.8±0.65µs        ? ?/sec    1.04    113.7±1.61µs        ? ?/sec
arrow_array_reader/BYTE_STREAM_SPLIT/Float16Array/byte_stream_split encoded, optional, no NULLs         1.00     56.7±1.02µs        ? ?/sec    1.04     59.0±0.52µs        ? ?/sec
```

**Comment by ShiKaiWi** [comment]
@Xuanwo @etseidl @alamb Existing benchmarks includes the (de)compression, and I run it with the current `unsafe` resize implementation and the benchmark shows a big improvement in decompression compared with the master branch:
```
compress SNAPPY - alphanumeric
                        time:   [54.489 µs 54.591 µs 54.699 µs]
                        change: [-35.659% -35.356% -35.055%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe

SNAPPY compressed 1048576 bytes of alphanumeric to 1048627 bytes
decompress SNAPPY - alphanumeric
                        time:   [45.716 µs 46.030 µs 46.348 µs]
                        change: [-34.833% -34.297% -33.757%] (p = 0.00 < 0.05)
                        Performance has improved.
```

PS: I run the command `cargo bench --bench compression --features="experimental default" -- SNAPPY`  under the directory `parquet`.

**Comment by ShiKaiWi** [comment]
The strict benchmark shows `prt::write_byte` + `Vec::set_len` has no performance improvement over the `Vec::resize` (actually, the underlying implement is just repeating `ptr::write_byte` and then `set_len`).

And to make a conclusion here, there are two implementations:
- `resize_without_init`: just use `Vec::set_len` to leave some contents in the buffer uninitialized, which is UB.
- `fast_resize`: `prt::write_bytes` and `Vec::set_len`

`resize_without_init` really performs better because the initialization work is skipped, but it breaks the safety contract of `Vec::set_len`.
And `fast_resize` does exactly the same work as `resize`.

I'm going to close this PR. Thanks a lot for guy's advice. @tustvold @alamb @Xuanwo @etseidl

**Comment by etseidl** [comment]
Thanks for your effort @ShiKaiWi. It got me to explore more of the `MaybeUninit` API.

@tustvold @alamb what is your philosophy when it comes to optimizations such as this, where unsafe code is used to gain some performance? I often find myself reaching for unsafe code to get a few more %, but wonder if that is generally frowned upon.

**Comment by alamb** [comment]
> Thanks for your effort @ShiKaiWi. 

I agree -- thank you @ShiKaiWi

**Comment by tustvold** [comment]
I'd echo Andrew's points and perhaps expand that this crate historically made a lot more use of unsafe. A significant effort over the last few years has been reducing this. This has not only yielded easier to maintain code, but in some cases also faster, as the compiler can better understand what is going on.

An example of where unsafe is easy to justify is the arithmetic kernels where the bounds checks otherwise break autovectorisation leading to a 8x performance hit. But if we're talking 5%, I'd probably think it isn't worth it

**Comment by Xuanwo** [comment]
Writing into uninitialized memory is a good optimization to me in this case. It has been a topic of discussion within the Rust community for years, with considerable effort dedicated to this area.

I recommend exploring this area after [BorrowedCursor](https://doc.rust-lang.org/std/io/struct.BorrowedCursor.html) has been stabilized or trying again with [`tokio::io::ReadBuf`](https://docs.rs/tokio/latest/tokio/io/struct.ReadBuf.html).

**Pull request closed without merging**